lib/puppet/functions/each.rb

```diff
@@ -1,38 +1,76 @@
+# value in turn to the lambda, then returns the original values.
 #
-#       [1,2,3].each |$val| { ... }                       # 1, 2, 3
-#       [5,6,7].each |$index, $val| { ... }               # (0, 5), (1, 6), (2, 7)
-#       {a=>1, b=>2, c=>3}].each |$val| { ... }           # ['a', 1], ['b', 2], ['c', 3]
-#       {a=>1, b=>2, c=>3}.each |$key, $val| { ... }      # ('a', 1), ('b', 2), ('c', 3)
-#       Integer[ 10, 20 ].each |$index, $value| { ... }   # (0, 10), (1, 11) ...
-#       "hello".each |$char| { ... }                      # 'h', 'e', 'l', 'l', 'o'
-#       3.each |$number| { ... }                          # 0, 1, 2
+# @example Using the `each` function with an array and a one-parameter lambda
+#
+# ~~~ puppet
+# # For the array $data, run a lambda outputting each value multiplied by 10 using the
+# # notice function, then return the array
+# $data = [1, 2, 3]
+# $each_data = $data.each |$item| { notice ($item * 10) }
+# # The notice function runs three times, outputting 10, then 20, then 30
+# # $each_data contains [1, 2, 3]
+# ~~~
+#
+# When the first argument is a hash, Puppet passes each key and value pair to the lambda
+# as an array in the form `[key, value]` and returns a hash containing the results.
+#
+# @example Using the `filter` function with a hash and a one-parameter lambda
+#
+# ~~~ puppet
+# # For the hash $data, run a lambda using each item as a key-value array and outputting
+# # the keys using the notice function, then return the hash
+# $data = {"a" => 0, "b" => 1, "c" => 2}
+# $each_data = $data.each |$items| { notice ($items[0]) }
+# # The notice function runs three times, outputting a, then b, then c
+# # $each_data contains {a => 0, b => 1, c => 2}
+# ~~~
+#
+# When the first argument is an array and the lambda has two parameters, Puppet passes the
+# array's indexes (enumerated from 0) in the first parameter and its values in the second
+# parameter.
+#
+# @example Using the `each` function with an array and a two-parameter lambda
+#
+# ~~~ puppet
+# # For the array $data, run a lambda outputting each index and its corresponding value,
+# # then return the array
+# $data = ["a", "b", "c"]
+# $each_data = $data.each |$index, $value| { notice ($index, $value) }
+# # The notice function runs three times, outputting 0 a, then 1 b, then 2 c
+# # $each_data contains [a, b, c]
+# ~~~
+#
+# When the first argument is a hash, Puppet passes its keys to the first parameter and its
+# values to the second parameter.
+#
+# @example Using the `each` function with a hash and a two-parameter lambda
+#
+# ~~~ puppet
+# # For the hash $data, run a lambda outputting each key and its corresponding value, then
+# # return the hash
+# $data = {"a" => 0, "b" => 1, "c" => 2}
+# $each_data = $data.each |$key, $value| { notice ($key, $value) }
+# # The notice function runs three times, outputting a 0, then b 1, then c 2
+# # $each_data contains {a => 0, b => 1, c => 2}
+# ~~~
```

Comment:
The 2-parameter examples should probably be the same task as the 1-parameter ones, since that gives the best parallax.